### PR TITLE
kubeadm: Refactor the .Etcd substruct in the v1alpha2 API

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:
Splits the monolithic `.Etcd` struct with all the options as fields to a more modular and clear design with two sub-structs for the different modes of hosting etcd we support.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of kubernetes/community#2131

**Special notes for your reviewer**:
Depends on:
 - [x] https://github.com/kubernetes/kubernetes/pull/63917

Follows up: https://github.com/kubernetes/kubernetes/pull/63871
TODO: I still need to write unit tests for this.

**Release note**:

```release-note
[action required] kubeadm: The `:Etcd` struct has been refactored in the v1alpha2 API. All the options now reside under either `.Etcd.Local` or `.Etcd.External`. Automatic conversions from the v1alpha1 API are supported.
```
@kubernetes/sig-cluster-lifecycle-pr-reviews @liztio

### CODE REVIEW

**Review by dixudx — COMMENTED** [review]
The old configurations for `Etcd` are replaced by `Etcd.Local` and `Etcd.External` respectively. The is a big change, but makes the configurations more concise.

Better add a notice somewhere to warn the users, since this new change will block in-place updates with old config file.

**Comment by luxas** [comment]
@timothysc this is ready to merge.

> Better add a notice somewhere to warn the users, since this new change will block in-place updates with old config file.

@dixudx we support full backwards-compability. All existing users will be upgraded to the new config automatically, thanks to the automatic API machinery conversations from `v1alpha1` to `v1alpha2`. Also, a release note with action required is automatically created from this PR.

**Comment by timothysc** [comment]
/assign @detiber @stealthybox  

They are the last two to update this area, they should have a say.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 1]
```diff
@@ -253,6 +252,54 @@ func ValidateTokenUsages(usages []string, fldPath *field.Path) field.ErrorList {
...
+		allErrs = append(allErrs, field.Invalid(fldPath, "", ".Etcd.Local and .Etcd.External are mutually exclusive"))
+		return allErrs
+	}
+	if e.Local != nil {
+		allErrs = append(allErrs, ValidateAbsolutePath(e.Local.DataDir, localPath.Child("dataDir"))...)
+		allErrs = append(allErrs, ValidateCertSANs(e.Local.ServerCertSANs, localPath.Child("serverCertSANs"))...)
+		allErrs = append(allErrs, ValidateCertSANs(e.Local.PeerCertSANs, localPath.Child("peerCertSANs"))...)
+	}
+	if e.External != nil {
+		allowHTTP := false
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: I think that `requireHTTPS` is more clear (the logic should be changed accordingly)

File: cmd/kubeadm/app/phases/certs/certs.go [hunk 2]
```diff
@@ -51,8 +51,7 @@ func CreatePKIAssets(cfg *kubeadmapi.MasterConfiguration) error {
 		CreateAPIServerEtcdClientCertAndKeyFiles,
 	}
 
-	// Currently this is the only way we have to identify static pod etcd vs external etcd
-	if len(cfg.Etcd.Endpoints) == 0 {
+	if cfg.Etcd.Local != nil || cfg.Etcd.External == nil {
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/phases/certs/certs.go]: `|| cfg.Etcd.External == nil`

**Review by fabriziopandini — COMMENTED** [review]
+💯 for cleaning up the etcd configuration
Only two minor comments from my side

File: cmd/kubeadm/app/phases/certs/certs.go (hunk 2 above)
Comment (luxas) [reply cmd/kubeadm/app/phases/certs/certs.go]: There is an edge case here actually when we don't know which one of the structs are defined or the other undefined (in unit tests), so hence I added this. It doesn't hurt I guess...?

**Comment by luxas** [comment]
> @luxas should we bump the api version for the featureGates struct and regenerate the config conversions before we merge this so that we have an independent patch for fixing the previous config issue we introduced with v1.10.0 ?

I've worked a lot the latest week on building on top of @liztio's initial great patch for that in order to make this all work flawlessly. See the earlier PRs linked in https://github.com/kubernetes/community/pull/2131

Basically what happens can be read in https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/config/masterconfig.go#L109-L160.
TL;DR; The featureGates and lack of TypeMeta when marshalled issues are resolved for the v1alpha1 version. That's the first that happens. Internally, when loading a v1alpha1 version in kubeadm v1.11, it will be automatically converted to the internal version which equals v1alpha2 which is the output format.

Does that answer your question?
I'm rebasing now and addressing @fabriziopandini's nits so we can merge.

File: cmd/kubeadm/app/phases/certs/certs.go (hunk 2 above)
Comment (fabriziopandini) [reply cmd/kubeadm/app/phases/certs/certs.go]: Ok. If possible this should be annotated in comments because this is the only place where this is required

**Comment by stealthybox** [comment]
/test pull-kubernetes-bazel-test
--
@luxas that does answer my question -- I'm glad the serialization issues are remedied.

File: cmd/kubeadm/app/phases/certs/certs.go (hunk 2 above)
Comment (luxas) [reply cmd/kubeadm/app/phases/certs/certs.go]: Okay, now I changed the unit tests instead, so now it works as you wanted it :+1:
